src/test/java/BaseTest.java

```diff
@@ -1,11 +1,81 @@
 import io.github.bonigarcia.wdm.WebDriverManager;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.chrome.ChromeDriver;
 import org.openqa.selenium.chrome.ChromeOptions;
-import org.testng.annotations.BeforeSuite;
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.testng.Assert;
+import org.testng.annotations.*;
+
+import java.time.Duration;
 
 public class BaseTest {
+    public WebDriver driver;
+
+    public WebDriverWait wait;
+    public Actions actions;
+
+    //public String url = "https://qa.koel.app/";
 
     @BeforeSuite
     static void setupClass() {
         WebDriverManager.chromedriver().setup();
     }
-}
\ No newline at end of file
+
+    @BeforeMethod
+    @Parameters({"BaseURL"})
+    public void launchBrowser(String baseURL) {
+        ChromeOptions options = new ChromeOptions();
+        options.addArguments("--remote-allow-origins=*");
+
+        driver = new ChromeDriver(options);
+        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
+        //Explicit Wait
+        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
+        actions = new Actions(driver);
+        driver.manage().window().maximize();
+        driver.get(baseURL);
+    }
+
+    @AfterMethod
+    public void closeBrowser() {
+
+        driver.quit();
+    }
+
+    public void navigateToPage(String url) {
+
+        driver.get(url);
+    }
+
+    public void logintoKoel() {
+        //WebElement loginBtn = driver.findElement(By.cssSelector("button[type='submit']"));
+        WebElement loginBtn = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("button[type='submit']")));
+        loginBtn.click();
+    }
+
+    public void provideEmail(String email) {
+        //WebElement emailField = driver.findElement(By.cssSelector("input[type='email']"));
+        WebElement emailField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='email']")));
+        emailField.clear();
+        emailField.sendKeys(email);
+    }
+
+    public void providePassword(String password) {
+        //WebElement passwordField = driver.findElement(By.cssSelector("input[type='password']"));
+        WebElement passwordField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='password']")));
+        passwordField.clear();
+        passwordField.sendKeys(password);
+    }
+
+    public void clickSubmit() {
+        WebElement submit = driver.findElement(By.cssSelector("button[type='submit']"));
+        submit.click();
+    }
+}
```

Comment:
**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://user-images.githubusercontent.com/104761482/233972627-2d7ffdc7-6a2f-4e9f-8a3a-a1b4a7a47bc4.png)

**Move these methods to the corresponding class according to the POM.**